### ISSUE TITLE
predicted values seem a bit off

### ISSUE THREAD

**Issue by simonheb** [sergiocorreia/reghdfe#137]
I'm using to predict but find something I consider unexpected, the fitted values seem to not exactly incorporate the fixed effects. The following minimal working example illustrates my point

```
clear 
set obs 100
gen x = rnormal()
gen d = mod(_n,2)
gen y=x+d

reghdfe y x , a(d, savefe)  
predict yhat, xbd //this is unexpected

reg y x i.d 
predict yhat2 //this has the expected behaviour

scatter yhat yhat2 y 

```
Since there is no uncertainty, the fitted values should be exactly recover the original y's, the standard `reg y x i.d` does what I expect, reghdfe doesn't. Am I using predict wrong here?

**Comment by sergiocorreia** [comment]
Hi Simon,

It looks like you have stumbled on a very odd bug from the old version of reghdfe (reghdfe versions from mid-2016 onwards shouldn't have this issue, but the SSC version is from early 2016).

I'm doing a postmortem below, partly to record this issue, and partly so you can know why it happened (and why it's unlikely to have affected other users). To follow, you need the latest versions of reghdfe and ftools (from github):

1) The problem is due to the fixed effects being incorrect, as show here:

```
clear 
set obs 10
set seed 123456
gen double x = rnormal()
gen double d = mod(_n,2)
gen double y = x + d
sort d y

reghdfe y x, a(FE1=d) old 
reghdfe y x, a(FE2=d)
gen double diff = FE2 - FE1
br y x d FE1 FE2 diff
```

2) The fixed effects are incorrect because the old version of reghdfe incorrectly reported `e(df_m)` as zero instead of 1. In that case, [line 2269](https://github.com/sergiocorreia/reghdfe/blob/master/src/reghdfe_old.ado#L2269) was executed, instead of line 2266. To prove this, see how the difference between the fixed effects is exactly the negative of the fixed effect of X:

```
clear 
set obs 10
set seed 123456
gen double x = rnormal()
gen double d = mod(_n,2)
gen double y = x + d
sort d y

reghdfe y x, a(FE1=d) old 
reghdfe y x, a(FE2=d)
gen double diff = FE2 - FE1

reghdfe x, a(d) resid
predict d_x, d
assert abs(diff + d_x) < 1e-10
br y x d FE1 FE2 diff d_x
```

3) Finally, the real bug, which was in the wrong `e(df_m)` (the degrees of freedom lost due to the Xs)

In [this line](https://github.com/sergiocorreia/reghdfe/blob/master/src/reghdfe_old.ado#L3295), we run Stata's `test` to get e(df_m). However, this doesn't work if the regression is perfectly explained (you can check it by running `areg y x, a(d)` and then `test x`)

Going back to the first example, notice how it works if we add some small error component to `y`:

```
clear 
set obs 10
set seed 123456
gen double x = rnormal()
gen double d = mod(_n,2)
gen double e = rnormal() / 100
gen double y = x + d + e
sort d y

reghdfe y x, a(FE1=d) old 
reghdfe y x, a(FE2=d)
gen double diff = FE2 - FE1
assert abs(diff) < 1e-10
```

---------------------

So, to recap, it seems that `predict,d` and `predict,xbd` gives you wrong results if these conditions hold:

- Older version of reghdfe (v3 or older)
- LHS variable is perfectly explained by the regressors

**Comment by simonheb** [comment]
Great, quick response. Indeed, updating as you suggested already solved the problem.
